### ISSUE TITLE
About the class probability field and semantic loss

### ISSUE THREAD

**Issue by XuYunqiu** [nihalsid/panoptic-lifting#19]
Hi, thanks for the awesome work. After reading the paper and code, I am confused about the class probability field and the semantic loss.

Different from prior method, the proposed method renders probabilities rather than logits. However, I find the `softmax` are used two times during calculating the semantics loss. 

1. The first `softmax` is used in the render to transfer the logits to the class probabilities:
https://github.com/nihalsid/panoptic-lifting/blob/f72fbfa1418638fef15a1c6e2b2ec8a0293871f0/model/renderer/panopli_tensoRF_renderer.py#L137-L139

2. The second `softmax` is included in the `torch.nn.CrossEntropyLoss` (i.e., `self.loss_semantics`):
https://github.com/nihalsid/panoptic-lifting/blob/f72fbfa1418638fef15a1c6e2b2ec8a0293871f0/trainer/train_panopli_tensorf.py#L145-L150

Should I change  the `torch.nn.CrossEntropyLoss` to `torch.nn.NLLLoss` in this line to avoid duplicate `softmax`?
https://github.com/nihalsid/panoptic-lifting/blob/f72fbfa1418638fef15a1c6e2b2ec8a0293871f0/trainer/train_panopli_tensorf.py#L64

**Comment by nihalsid** [comment]
Hi, thank you for your interest in the work!

In the "softmax" mode, the semantic branch outputs probabilities using softmax:
https://github.com/nihalsid/panoptic-lifting/blob/f72fbfa1418638fef15a1c6e2b2ec8a0293871f0/trainer/train_panopli_tensorf.py#L51C11-L51C130

However since the crossentropy loss expects unnormalized logits instead of probabilities, we convert it into scores using:
https://github.com/nihalsid/panoptic-lifting/blob/f72fbfa1418638fef15a1c6e2b2ec8a0293871f0/model/renderer/panopli_tensoRF_renderer.py#L139

Note that a proper probability -> logit conversion would have been log(p/(1-p)) but we found the above formulation for scores to work better.

**Comment by XuYunqiu** [comment]
I see. Thanks again for your quick reply.

**Comment by XuYunqiu** [comment]
Hi @nihalsid, I have one more question about the segment clustering loss. 
I'm wondering why it only uses rays from one image instead of sampling rays across the scenes.
Looking forward to your reply. Thanks!

**Comment by nihalsid** [comment]
Hi Yunqiu,

Since the segment IDs are valid only per frame, we only sample rays in the same frame. 

E.g. the segment id 1 in frame 1 is not the same as segment id 1 in frame 2.

For the same reason, instances are also dealt in the same way.

Thanks!

**Comment by XuYunqiu** [comment]
In my understanding, for the semantic segmentation case, we can get the semantic classes (i.e., pseudo semantic masks) among different frames.
Can we sample the rays from different frames based on such pseudo labels?

**Comment by nihalsid** [comment]
Semantic segmentations is sampled across frames already, its the segment clustering loss that is done per frame. There's a difference between the two -- semantic segmentation is class dependent, but segment independent, segment clustering loss is segment dependent but class independent.

**Comment by XuYunqiu** [comment]
Got it. The segment clustering loss is based on the panoptic segments (including thing instances and stuff segments).
Thanks for your prompt and kind response.

**Comment by nihalsid** [comment]
No problem! :)
